File: src/main/java/com/nobility/downloader/MainController.java
package com.nobility.downloader;

import com.nobility.downloader.downloads.Download;
import com.nobility.downloader.settings.Defaults;
import com.nobility.downloader.utils.AlertBox;
import com.nobility.downloader.utils.StringChecker;
import com.nobility.downloader.utils.Toast;
import com.nobility.downloader.utils.Tools;
import javafx.beans.property.SimpleStringProperty;
import javafx.event.ActionEvent;
import javafx.fxml.FXML;
import javafx.fxml.FXMLLoader;
import javafx.fxml.Initializable;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.control.Button;
import javafx.scene.control.MenuItem;
import javafx.scene.control.TextArea;
import javafx.scene.control.TextField;
import javafx.scene.control.*;
import javafx.scene.control.cell.PropertyValueFactory;
import javafx.scene.image.Image;
import javafx.scene.input.KeyCode;
import javafx.stage.Modality;
import javafx.stage.Stage;
import javafx.stage.StageStyle;

import java.awt.*;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.lang.reflect.Constructor;
import java.net.URL;
import java.text.SimpleDateFormat;
import java.util.Comparator;
import java.util.Date;
import java.util.ResourceBundle;

public class MainController implements Initializable {

    private final Model model;
    @FXML
    public TextField tf_sub_url;
    @FXML
    public TextArea print;
    @FXML
    public MenuItem open_download_folder, about, open_settings, updates, open_download_history, open_website, open_github;
    @FXML
    private Button stop_button, start_button;
    @FXML
    private TableView<Download> download_table;
    @FXML
    private TableColumn<Download, String> name_column, size_column, progress_column, date_column;

    public MainController(Model model) {
        this.model = model;
        //SeriesScaper seriesScaper = new SeriesScaper(model);
        //seriesScaper.scrapeSeriesIntoFile();
    }

    public void setMainStage(Stage mainStage) {
        model.setMainStage(mainStage);
        mainStage.widthProperty().addListener(((observable, oldValue, newValue) -> download_table.refresh()));
    }

    @Override
    public void initialize(URL location, ResourceBundle resources) {
        model.setTextOutput(print);
        model.setUrlTextField(tf_sub_url);
        tf_sub_url.promptTextProperty().setValue(Model.EXAMPLE_SERIES);
        tf_sub_url.setOnKeyPressed(e -> {
            if (e.getCode().equals(KeyCode.ENTER)) {
                start();
            }
        });
        stop_button.setDisable(true);
        model.setStartButton(start_button);
        model.setStopButton(stop_button);
        String lastUrl = model.settings().getString(Defaults.LASTDOWNLOAD);
        if (!StringChecker.isNullOrEmpty(lastUrl)) {
            tf_sub_url.setText(lastUrl);
        }
        download_table.setColumnResizePolicy(TableView.CONSTRAINED_RESIZE_POLICY);
        name_column.setMaxWidth(1f * Integer.MAX_VALUE * 60);
        date_column.setMaxWidth(1f * Integer.MAX_VALUE * 20);
        size_column.setMaxWidth(1f * Integer.MAX_VALUE * 10);
        progress_column.setMaxWidth(1f * Integer.MAX_VALUE * 7);

        name_column.setCellValueFactory(new PropertyValueFactory<>("name"));
        size_column.setCellValueFactory(row -> new SimpleStringProperty(Tools.bytesToString(row
                .getValue().getFileSize())));
        progress_column.setCellValueFactory(row -> row.getValue().getProgress());
        date_column.setCellValueFactory(new PropertyValueFactory<>("dateAdded"));
        date_column.setSortType(TableColumn.SortType.DESCENDING);
        download_table.getSortOrder().add(date_column);
        Comparator<String> dateComparator = (o1, o2) -> {
            SimpleDateFormat sdf = new SimpleDateFormat(Tools.dateFormat);
            try {
                Date date1 = sdf.parse(o1);
                Date date2 = sdf.parse(o2);
                return date1.compareTo(date2);
            } catch (Exception ignored) {
            }
            return 0;
        };
        date_column.setComparator(dateComparator);

        download_table.setRowFactory(param -> {
            TableRow<Download> row = new TableRow<>();
            row.emptyProperty().addListener((observable, oldValue, isEmpty) -> {
                if (!isEmpty) {
                    Download download = row.getItem();
                    if (download != null) {
                        ContextMenu menu = new ContextMenu();
                        MenuItem openFolder = new MenuItem("Open Folder");
                        openFolder.setOnAction(event -> model.openFolder(download.getDownloadPath(), true));
                        MenuItem deleteFile = new MenuItem("Delete File");
                        deleteFile.setOnAction(event ->
                                model.showConfirm("Do you wish to delete this file?", () -> {
                                    if (download.isDownloading()) {
                                        Toast.makeToast(model.getMainStage(), "You can't delete videos that are being downloaded.");
                                        return;
                                    }
                                    File file = new File(download.getDownloadPath());
                                    if (file.exists()) {
                                        if (file.delete()) {
                                            model.showMessage("Success", "Successfully deleted this episode.");
                                        } else {
                                            model.showError("Unable to delete this file. No error thrown. Most likely folder permission issues.");
                                        }
                                    } else {
                                        Toast.makeToast(model.getMainStage(), "The original file no longer exists.");
                                    }
                                }));
                        MenuItem removeFromList = new MenuItem("Remove From List");
                        removeFromList.setOnAction(event -> {
                            if (download.isDownloading()) {
                                Toast.makeToast(model.getMainStage(),"You can't remove downloads that are being downloaded.");
                                return;
                            }
                            model.removeDownload(download);
                        });

                        MenuItem resumeDownload = new MenuItem("Resume Download");
                        resumeDownload.setOnAction(event -> {
                            if (download.isComplete()) {
                                Toast.makeToast(model.getMainStage(),"This download is completed. You can't resume it.");
                                return;
                            }
                            if (model.isRunning()) {
                                if (model.getLinkUrls().contains(download.getUrl())) {
                                    Toast.makeToast(model.getMainStage(), "This download is already in queue.");
                                    return;
                                }
                                model.getLinks().add(download.toEpisode());
                                Toast.makeToast(model.getMainStage(), "Successfully added " + download.getName() + " to current queue.");
                            } else {
                                tf_sub_url.setText(download.getUrl());
                                start();
                                Toast.makeToast(model.getMainStage(), "Launched video downloader for: " + download.getName());
                            }
                        });
                        MenuItem openDownloadUrl = new MenuItem("Open Download URL");
                        openDownloadUrl.setOnAction(event -> model.showLinkPrompt(download.getUrl(), true));
                        MenuItem copyDownloadUrl = new MenuItem("Copy Download URL");
                        copyDownloadUrl.setOnAction(event -> model.showCopyPrompt(download.getUrl(), false));
                        MenuItem seriesDetails = new MenuItem("Series Details");
                        seriesDetails.setOnAction(event -> model.openVideoDetails(download.getSeriesLink()));
                        MenuItem copySeriesLink = new MenuItem("Copy Series Lihk");
                        copySeriesLink.setOnAction(event -> model.showCopyPrompt(download.getSeriesLink(), false));
                        MenuItem pauseDownload = new MenuItem("Pause Download");
                        pauseDownload.setOnAction(event -> {
                            //
                            Toast.makeToast(model.getMainStage(), "Not implemented.");
                        });
                        MenuItem play = new MenuItem("Play Video");
                        play.setOnAction(event -> {
                            try {
                                if (Desktop.isDesktopSupported()) {
                                    Desktop.getDesktop().open(download.getDownloadFile());
                                } else {
                                    Toast.makeToast(model.getMainStage(), "Desktop is not supported.");
                                }
                            } catch (Exception e) {
                                Toast.makeToast(model.getMainStage(), "Failed to play video. Check the console.");
                                if (e.getLocalizedMessage().contains("No application is associated with the specified file for this operation")) {
                                    System.out.println("There is no default application for opening this type of file. (mp4)");
                                } else {
                                    System.out.println(e.getLocalizedMessage());
                                }
                            }
                        });
                        menu.getItems().addAll(
                                openFolder,
                                openDownloadUrl,
                                copyDownloadUrl
                        );
                        if (!StringChecker.isNullOrEmpty(download.getSeriesLink())) {
                            menu.getItems().add(seriesDetails);
                            menu.getItems().add(copySeriesLink);
                        }
                        if (download.isDownloading() && !download.isComplete()) {
                            menu.getItems().add(0, pauseDownload);
                        }
                        if (!download.isDownloading() && !download.isComplete()) {
                            menu.getItems().add(0, resumeDownload);
                        }
                        if (!download.isDownloading() && download.isComplete()) {
                            menu.getItems().add(0, play);
                        }
                        if (!download.isDownloading()) {
                            menu.getItems().add(removeFromList);
                        }
                        if (download.isComplete() && !download.isDownloading()) {
                            menu.getItems().add(deleteFile);
                        }
                        row.setContextMenu(menu);

                        row.setOnMouseClicked(event -> {
                            if (model.settings().getBoolean(Defaults.SHOWCONTEXTONCLICK)) {
                                menu.show(model.getMainStage(), event.getScreenX(), event.getScreenY());
                            }
                        });

                    }
                }
            });
            return row;
        });
        model.setTableView(download_table);
        model.getUpdateManager().checkForUpdates(false, true);
    }

    private Stage settingsStage;

    public void openSettings() {
        if (settingsStage == null) {
            settingsStage = new Stage();
            settingsStage.initModality(Modality.APPLICATION_MODAL);
            settingsStage.initOwner(model.getMainStage());
            settingsStage.setTitle("Settings");
            InputStream icon = Main.class.getResourceAsStream(Model.SETTINGS_ICON);
            if (icon != null) {
                settingsStage.getIcons().add(new Image(icon));
            }
            settingsStage.setResizable(true);
            settingsStage.initStyle(StageStyle.DECORATED);
        }
        FXMLLoader loader = new FXMLLoader(Main.class.getResource(Model.FX_PATH + "settings.fxml"));
        loader.setControllerFactory((Class<?> controllerType) -> {
            try {
                for (Constructor<?> con : controllerType.getConstructors()) {
                    if (con.getParameterCount() == 2 && con.getParameterTypes()[0] == Model.class
                            && con.getParameterTypes()[1] == Stage.class) {
                        return con.newInstance(model, settingsStage);
                    }
                }
                return controllerType.getDeclaredConstructor().newInstance();
            } catch (Exception e) {
                System.err.println("Failed to load SettingsController. Error: " + e.getMessage());
                e.printStackTrace(System.err);
                return null;
            }
        });
        Parent layout;
        try {
            layout = loader.load();
            Scene scene = new Scene(layout);
            settingsStage.toFront();
            settingsStage.setScene(scene);
            settingsStage.sizeToScene();
            settingsStage.showAndWait();
        } catch (IOException e) {
            e.printStackTrace();
            System.out.println("Validation Error: " + e.getMessage());
        }
    }

    private Stage historyStage;

    private void openHistory() {
        if (historyStage == null) {
            historyStage = new Stage();
            historyStage.initModality(Modality.APPLICATION_MODAL);
            historyStage.initOwner(model.getMainStage());
            historyStage.setTitle("Series Download History");
            InputStream icon = Main.class.getResourceAsStream(Model.SETTINGS_ICON);
            if (icon != null) {
                historyStage.getIcons().add(new Image(icon));
            }
            historyStage.setResizable(true);
            historyStage.initStyle(StageStyle.DECORATED);
        }
        FXMLLoader loader = new FXMLLoader(Main.class.getResource(Model.FX_PATH + "history.fxml"));
        loader.setControllerFactory((Class<?> controllerType) -> {
            try {
                for (Constructor<?> con : controllerType.getConstructors()) {
                    if (con.getParameterCount() == 2 && con.getParameterTypes()[0] == Model.class) {
                        return con.newInstance(model, historyStage);
                    }
                }
                return controllerType.getDeclaredConstructor().newInstance();
            } catch (Exception e) {
                System.err.println("Failed to load HistoryController. Error: " + e.getMessage());
                e.printStackTrace(System.err);
                return null;
            }
        });
        Parent layout;
        try {
            layout = loader.load();
            Scene scene = new Scene(layout);
            scene.getStylesheets().add(String.valueOf(Main.class.getResource(Model.CSS_PATH + "contextmenu.css")));
            historyStage.toFront();
            historyStage.setScene(scene);
            historyStage.sizeToScene();
            historyStage.showAndWait();
        } catch (IOException e) {
            e.printStackTrace();
            System.out.println("Validation Error: " + e.getMessage());
        }
    }

    @FXML
    private void clearConsole() {
        print.clear();
    }

    @FXML
    private void start() {
        model.start();
    }

    @FXML
    private void stop() {
        model.stop();
    }

    @FXML
    public void handleClicks(ActionEvent event) throws IOException {
        Object src = event.getSource();
        if (src.equals(open_settings)) {
            openSettings();
        } else if (src.equals(open_download_history)) {
            openHistory();
        } else if (src.equals(open_download_folder)) {
            if (Desktop.isDesktopSupported()) {
                if (!StringChecker.isNullOrEmpty(model.settings().getString(Defaults.SAVEFOLDER))) {
                    File f = new File(model.settings().getString(Defaults.SAVEFOLDER));
                    Desktop.getDesktop().open(f);
                } else {
                    AlertBox.show(Alert.AlertType.ERROR, "Your download folder doesn't exist.", "Be sure to set it inside " +
                            "the settings before downloading videos.");
                }
            } else {
                model.showError("Desktop is not supported on your OS.");
            }
        } else if (src.equals(about)) {
            model.showMessage("About", "This is a FREE open source program to download videos from " + Model.WEBSITE
                    + ". That's all. :) "
                    + "\nAuthor Discord: Nobility#9814"
                    + "\nGithub: https://github.com/NobilityDeviant/"
                    + "\nTo use this program you must first install Google Chrome. Everything else should be " +
                    "handled automatically." +
                    "\nYou find an episode or series link at " + Model.WEBSITE + " and paste it into " +
                    "the field. The default settings should be enough for most people." +
                    "\nIf you have any issues, please let me know on Discord.");
        } else if (src.equals(updates)) {
            if (model.settings().getBoolean(Defaults.DENIEDUPDATE)) {
                model.settings().setBoolean(Defaults.DENIEDUPDATE, false);
                model.saveSettings();
            }
            model.getUpdateManager().checkForUpdates(true, true);
        } else if (src.equals(open_website)) {
            model.showLinkPrompt(Model.WEBSITE, true);
        } else if (src.equals(open_github)) {
            model.showLinkPrompt(Model.GITHUB, true);
        }
    }
}

File: src/main/java/com/nobility/downloader/MainController.kt
package com.nobility.downloader

import com.nobility.downloader.entities.Download
import com.nobility.downloader.settings.Defaults
import com.nobility.downloader.utils.Tools
import com.nobility.downloader.utils.Tools.fixOldLink
import javafx.beans.property.SimpleStringProperty
import javafx.beans.value.ObservableValue
import javafx.event.ActionEvent
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.fxml.FXMLLoader
import javafx.fxml.Initializable
import javafx.scene.Parent
import javafx.scene.Scene
import javafx.scene.control.*
import javafx.scene.image.Image
import javafx.scene.input.*
import javafx.stage.Stage
import javafx.util.Callback
import kotlinx.coroutines.launch
import java.awt.Desktop
import java.io.File
import java.io.IOException
import java.net.URL
import java.text.SimpleDateFormat
import java.util.*


class MainController(private val model: Model, private val mainStage: Stage) : Initializable {

    @FXML
    private lateinit var urlSubmission: TextField

    @FXML
    private lateinit var console: TextArea

    @FXML
    private lateinit var openDownloadFolder: MenuItem

    @FXML
    private lateinit var about: MenuItem

    @FXML
    private lateinit var openSettings: MenuItem

    @FXML
    private lateinit var updates: MenuItem

    @FXML
    private lateinit var openDownloadHistory: MenuItem

    @FXML
    private lateinit var openWcoSeries: MenuItem

    @FXML
    private lateinit var openWebsite: MenuItem

    @FXML
    private lateinit var openGithub: MenuItem

    @FXML
    private lateinit var stopButton: Button

    @FXML
    private lateinit var startButton: Button

    @FXML
    private lateinit var downloadTable: TableView<Download>

    @FXML
    private lateinit var nameColumn: TableColumn<Download, String>

    @FXML
    private lateinit var sizeColumn: TableColumn<Download, String>

    @FXML
    private lateinit var progressColumn: TableColumn<Download, String>

    @FXML
    private lateinit var dateColumn: TableColumn<Download, String>

    override fun initialize(location: URL, resources: ResourceBundle?) {
        model.setTextOutput(console)
        model.urlTextField = urlSubmission
        urlSubmission.promptTextProperty().value = Model.EXAMPLE_SERIES
        urlSubmission.onKeyPressed = EventHandler { e: KeyEvent ->
            if (e.code == KeyCode.ENTER) {
                start()
            }
        }
        stopButton.isDisable = true
        model.setStartButton(startButton)
        model.setStopButton(stopButton)
        val lastUrl = model.settings().stringSetting(Defaults.LASTDOWNLOAD)
        if (lastUrl.isNotEmpty()) {
            urlSubmission.text = lastUrl
        }
        downloadTable.columnResizePolicy = TableView.CONSTRAINED_RESIZE_POLICY
        nameColumn.maxWidth = (1f * Int.MAX_VALUE * 60).toDouble()
        dateColumn.maxWidth = (1f * Int.MAX_VALUE * 20).toDouble()
        sizeColumn.maxWidth = (1f * Int.MAX_VALUE * 10).toDouble()
        progressColumn.maxWidth = (1f * Int.MAX_VALUE * 7).toDouble()
        nameColumn.setCellValueFactory {
            SimpleStringProperty(it.value.name)
        }
        sizeColumn.setCellValueFactory {
            it.value.fileSizeProperty
        }
        progressColumn.setCellValueFactory {
            it.value.progress
        }
        dateColumn.setCellValueFactory {
            it.value.dateProperty
        }
        dateColumn.sortType = TableColumn.SortType.DESCENDING
        val dateComparator = Comparator { o1: String?, o2: String? ->
            val sdf = SimpleDateFormat(Tools.dateFormat)
            try {
                val date1 = sdf.parse(o1)
                val date2 = sdf.parse(o2)
                return@Comparator date1.compareTo(date2)
            } catch (ignored: Exception) {
            }
            0
        }
        dateColumn.comparator = dateComparator
        downloadTable.setRowFactory {
            val row = TableRow<Download>()
            row.isCache = false
            row.emptyProperty()
                .addListener { _: ObservableValue<out Boolean?>?, _: Boolean?, isEmpty: Boolean? ->
                    if (isEmpty == false) {
                        if (row.item != null) {
                            val menu = ContextMenu()
                            val openFolder = MenuItem("Open Folder")
                            openFolder.onAction =
                                EventHandler { model.openFolder(row.item.downloadPath, true) }
                            val deleteFile = MenuItem("Delete File")
                            deleteFile.onAction = EventHandler {
                                model.showConfirm("Do you wish to delete this file?") {
                                    if (row.item.downloading || row.item.queued) {
                                        model.toast("You can't delete videos that are being downloaded.")
                                        return@showConfirm
                                    }
                                    val file = row.item.downloadFile()
                                    if (file != null && file.exists()) {
                                        if (file.delete()) {
                                            model.removeDownload(row.item)
                                            model.showMessage("Success", "Successfully deleted this episode.")
                                        } else {
                                            model.showError("Unable to delete this file. No error thrown. Most likely folder permission issues.")
                                        }
                                    } else {
                                        model.toast("The original file no longer exists.")
                                    }
                                }
                            }
                            val removeFromList = MenuItem("Remove From List")
                            removeFromList.onAction = EventHandler {
                                if (row.item.downloading || row.item.queued) {
                                    model.toast("You can't remove downloads that are being downloaded.")
                                    return@EventHandler
                                }
                                model.removeDownload(row.item)
                            }
                            val resumeDownload = MenuItem("Resume Download")
                            resumeDownload.onAction = EventHandler {
                                if (row.item.isComplete) {
                                    model.toast("This download is completed. You can't resume it.")
                                    return@EventHandler
                                }
                                if (model.isRunning) {
                                    if (model.addEpisodeToQueue(row.item)) {
                                        model.toast("Successfully added episode to current queue.")
                                    } else {
                                        model.toast("This episode is already in queue.")
                                    }
                                } else {
                                    urlSubmission.text = fixOldLink(row.item.link)
                                    start()
                                    model.toast("Launched video downloader for: ${row.item.name}")
                                }
                            }
                            val openDownloadUrl = MenuItem("Open Download URL")
                            openDownloadUrl.onAction = EventHandler {
                                model.showLinkPrompt(
                                    fixOldLink(row.item.link), true
                                )
                            }
                            val copyDownloadUrl = MenuItem("Copy Download URL")
                            copyDownloadUrl.onAction = EventHandler {
                                model.showCopyPrompt(
                                    fixOldLink(row.item.link), false
                                )
                            }
                            val seriesDetails = MenuItem("Series Details")
                            seriesDetails.onAction = EventHandler {
                                model.openSeriesDetails(
                                    fixOldLink(row.item.seriesLink)
                                )
                            }
                            val copySeriesLink = MenuItem("Copy Series Lihk")
                            copySeriesLink.onAction = EventHandler {
                                model.showCopyPrompt(
                                    fixOldLink(row.item.seriesLink),
                                    false
                                )
                            }
                            val downloadSeries = MenuItem("Download Series")
                            downloadSeries.onAction = EventHandler {
                                if (model.isRunning) {
                                    model.showError("You can't download a series while the downloader is running.")
                                    return@EventHandler
                                }
                                model.urlTextField.text = row.item.seriesLink
                                model.start()
                                //model.toast(
                                  //  "Successfully launched video downloader for: ${row.item.seriesLink}"
                                //)
                            }
                            val pauseDownload = MenuItem("Pause Download")
                            pauseDownload.onAction = EventHandler {
                                model.toast("Not implemented.")
                            }
                            val play = MenuItem("Play Video")
                            play.onAction = EventHandler {
                                try {
                                    if (Desktop.isDesktopSupported()) {
                                        model.taskScope.launch {
                                            Desktop.getDesktop().open(row.item.downloadFile())
                                        }
                                    } else {
                                        model.toast("Desktop is not supported.")
                                    }
                                } catch (e: Exception) {
                                    model.toast("Failed to play video. Check the console.")
                                    if (e.localizedMessage.contains("No application is associated with the specified file for this operation")) {
                                        println("There is no default application for opening this type of file. (mp4)")
                                    } else {
                                        println(e.localizedMessage)
                                    }
                                }
                            }
                            menu.items.addAll(
                                openFolder,
                                openDownloadUrl,
                                copyDownloadUrl
                            )
                            if (!row.item.seriesLink.isNullOrEmpty()) {
                                menu.items.add(seriesDetails)
                                menu.items.add(copySeriesLink)
                                menu.items.add(downloadSeries)
                            }
                            if ((row.item.downloading || row.item.queued) && !row.item.isComplete) {
                                menu.items.add(0, pauseDownload)
                            }
                            if (!row.item.downloading &&!row.item.queued && !row.item.isComplete) {
                                menu.items.add(0, resumeDownload)
                            }
                            if (!row.item.downloading && !row.item.queued && row.item.isComplete) {
                                menu.items.add(0, play)
                            }
                            if (!row.item.downloading && !row.item.queued) {
                                menu.items.add(removeFromList)
                            }
                            if (row.item.isComplete && !row.item.downloading && !row.item.queued) {
                                menu.items.add(deleteFile)
                            }
                            row.contextMenu = menu
                            row.onMouseClicked = EventHandler { event: MouseEvent ->
                                if (model.settings().booleanSetting(Defaults.SHOWCONTEXTONCLICK)) {
                                    menu.show(model.mainStage, event.screenX, event.screenY)
                                }
                            }
                        }
                    }
                }
            row
        }
        model.setTableView(downloadTable, dateColumn)
        model.taskScope.launch {
            model.updateManager.checkForUpdates(prompt = false, refresh = true)
        }
        model.mainStage = mainStage
        mainStage.widthProperty()
            .addListener { _: ObservableValue<out Number?>?, _: Number?, _: Number? ->
                downloadTable.refresh()
            }
        val downloadFolder = File(model.settings().stringSetting(Defaults.SAVEFOLDER))
        if (!downloadFolder.exists()) {
            println("Your download folder no longer exists." +
                    "\n${downloadFolder.absolutePath}")
            println("Your download folder has been reset to the default.")
            model.settings().setSetting(Defaults.SAVEFOLDER, Defaults.SAVEFOLDER.value)
        }
        if (!downloadFolder.canWrite()) {
            println(
                """
    Your download folder doesn't allow write permissions.
    If this is a USB or SD Card then disable write protection.
    Try selecting a folder in the user or home folder. Those are usually not restricted.
    ${downloadFolder.absolutePath}
    """.trimIndent()
            )
            println("Your download folder has been reset to the default.")
            model.settings().setSetting(Defaults.SAVEFOLDER, Defaults.SAVEFOLDER.value)
        }
    }

    private var historyStage: Stage = Stage()

    private fun openHistory() {
        historyStage.title = "Series Download History"
        val icon = Main::class.java.getResourceAsStream(Model.SETTINGS_ICON)
        if (icon != null) {
            historyStage.icons.add(Image(icon))
        }
        historyStage.isResizable = true
        val loader = FXMLLoader(Main::class.java.getResource(Model.FX_PATH + "history.fxml"))
        loader.controllerFactory = Callback { controllerType: Class<*> ->
            try {
                for (con in controllerType.constructors) {
                    if (con.parameterCount == 2 && con.parameterTypes[0] == Model::class.java) {
                        return@Callback con.newInstance(model, historyStage)
                    }
                }
                return@Callback controllerType.getDeclaredConstructor().newInstance()
            } catch (e: Exception) {
                System.err.println("Failed to load HistoryController. Error: ${e.localizedMessage}")
                e.printStackTrace()
                return@Callback null
            }
        }
        val layout: Parent
        try {
            layout = loader.load()
            val scene = Scene(layout)
            scene.stylesheets.add(
                Main::class.java.getResource(
                    Model.CSS_PATH + "contextmenu.css"
                )?.toString() ?: ""
            )
            historyStage.toFront()
            historyStage.scene = scene
            historyStage.sizeToScene()
            historyStage.show()
        } catch (e: IOException) {
            e.printStackTrace()
            println("Failed to load history window. Error: ${e.localizedMessage}")
        }
    }

    @FXML
    fun clearURL() {
        urlSubmission.clear()
    }

    @FXML
    fun pasteURL() {
        val clipboard = Clipboard.getSystemClipboard()
        urlSubmission.text = clipboard.string
    }

    @FXML
    fun pasteURLAndStart() {
        val clipboard = Clipboard.getSystemClipboard()
        urlSubmission.text = clipboard.string
        if (!model.isRunning) {
            model.start()
        } else {
            //todo make a way to do this in the background without effecting the current dl
            //if found in cache just launch
            model.showError("Downloader is already running. Unable to start.")
        }
    }

    @FXML
    fun clearConsole() {
        console.clear()
    }

    @FXML
    fun start() {
        model.start()
    }

    @FXML
    fun stop() {
        model.stop()
    }

    fun setupHotKeys() {
        model.mainStage.scene.addEventFilter(KeyEvent.KEY_PRESSED, object : EventHandler<KeyEvent> {
            val settings: KeyCombination = KeyCodeCombination(
                KeyCode.S,
                KeyCombination.CONTROL_DOWN
            )
            val history: KeyCombination = KeyCodeCombination(
                KeyCode.H,
                KeyCombination.CONTROL_DOWN
            )
            val downloadFolder: KeyCombination = KeyCodeCombination(
                KeyCode.F,
                KeyCombination.CONTROL_DOWN
            )
            val wco: KeyCombination = KeyCodeCombination(
                KeyCode.W,
                KeyCombination.CONTROL_DOWN
            )


            override fun handle(ke: KeyEvent) {
                if (settings.match(ke)) {
                    model.openSettings()
                    ke.consume()
                } else if (wco.match(ke)) {
                    model.openWco()
                    ke.consume()
                } else if (history.match(ke)) {
                    openHistory()
                    ke.consume()
                } else if (downloadFolder.match(ke)) {
                    openDownloadFolder()
                    ke.consume()
                }
            }
        })
        //println("Hotkeys have been set up.")
    }

    private fun openDownloadFolder() {
        if (Desktop.isDesktopSupported()) {
            if (model.settings().stringSetting(Defaults.SAVEFOLDER).isNotEmpty()) {
                val f = File(model.settings().stringSetting(Defaults.SAVEFOLDER))
                Desktop.getDesktop().open(f)
            } else {
                model.showError(
                    "Your download folder doesn't exist.",
                    "Be sure to set it inside the settings before downloading videos."
                )
                model.openSettings(0)
            }
        } else {
            model.showError("Desktop is not supported on your OS.")
        }
    }

    @FXML
    @Throws(IOException::class)
    fun handleClicks(event: ActionEvent) {
        when (event.source) {
            openSettings -> {
                model.openSettings()
            }
            openDownloadHistory -> {
                openHistory()
            }
            openWcoSeries -> {
                model.openWco()
            }
            openDownloadFolder -> {
                openDownloadFolder()
            }
            about -> {
                model.showMessage(
                    "About",
                    "This is a FREE open source program to download videos from ${Model.WEBSITE}. That's all. :)" +
                            "\nAuthor Discord: I don't have discord anymore. Contact me on Github." +
                            "\nGithub: https://github.com/NobilityDeviant/" +
                            "\nTo use this program you must first install a browser like: " +
                            "\nGoogle Chrome, Chromium, Opera, Edge, Firefox or Safari." +
                            "\nYou can choose any listed browser in the settings. Everything else should be handled automatically." +
                            "\nYou find an episode or series link at ${Model.WEBSITE} and paste it into the field. The default settings should be enough for most people." +
                            "\nIf you have any issues, please create an issue on Github."
                )
            }
            updates -> {
                if (model.settings().booleanSetting(Defaults.DENIEDUPDATE)) {
                    model.settings().setSetting(Defaults.DENIEDUPDATE, false)
                }
                model.taskScope.launch {
                    model.updateManager.checkForUpdates(prompt = true, refresh = true)
                }
            }
            openWebsite -> {
                model.showLinkPrompt(Model.WEBSITE, true)
            }
            openGithub -> {
                model.showLinkPrompt(Model.GITHUB, true)
            }
        }
    }
}